File: framework/wcf/Extensibility/Transactions/TransactionMessagePropertyUdpTransport/VB/MessageInspectors/TransactionFlowProperty.vb
'  Copyright (c) Microsoft Corporation. All rights reserved.

Imports System.ServiceModel
Imports System.ServiceModel.Channels

Namespace Microsoft.ServiceModel.Samples
	Public Class TransactionFlowProperty
        Public Const PropertyName = "TransactionFlowProperty"
		Private propToken() As Byte

		Private Sub New()
		End Sub

		Public Shared Function [Get](ByVal message As Message) As Byte()
			If message Is Nothing Then
				Return Nothing
			End If

			If message.Properties.ContainsKey(PropertyName) Then
				Dim tfp As TransactionFlowProperty = CType(message.Properties(PropertyName), TransactionFlowProperty)
				Return tfp.propToken
			End If

			Return Nothing
		End Function

		Public Shared Sub [Set](ByVal propToken() As Byte, ByVal message As Message)
			If message.Properties.ContainsKey(PropertyName) Then
				Throw New CommunicationException("A transaction flow property is already set on the message.")
			End If

			Dim [property] As New TransactionFlowProperty()
			[property].propToken = propToken
			message.Properties.Add(PropertyName, [property])
		End Sub
	End Class
End Namespace

File: framework/wcf/Extensibility/Transactions/TransactionMessagePropertyUdpTransport/CS/MessageInspectors/TransactionFlowProperty.cs
//  Copyright (c) Microsoft Corporation. All rights reserved.

using System.ServiceModel;
using System.ServiceModel.Channels;

namespace Microsoft.ServiceModel.Samples
{
    public class TransactionFlowProperty
    {
        public const string PropertyName = "TransactionFlowProperty";
        byte[] propToken;

        private TransactionFlowProperty()
        {
        }

        public static byte[] Get(Message message)
        {
            if (message == null)
            {
                return null;
            }

            if (message.Properties.ContainsKey(PropertyName))
            {
                TransactionFlowProperty tfp = (TransactionFlowProperty)message.Properties[PropertyName]; 
                return tfp.propToken;
            }

            return null;
        }

        public static void Set(byte[] propToken, Message message)
        {
            if (message.Properties.ContainsKey(PropertyName))
            {
                throw new CommunicationException("A transaction flow property is already set on the message.");
            }

            TransactionFlowProperty property = new TransactionFlowProperty();
            property.propToken = propToken;
            message.Properties.Add(PropertyName, property);
        }
    }
}
